### ISSUE TITLE
array_map always set to return an array (even if a list is passed)

### ISSUE THREAD

**Issue by wouterj** [vimeo/psalm#4811]
https://psalm.dev/r/40b30bdcbd

As `array_map` only modifies the item value, I think passing a list should also be understood as returning a list.

I couldn't find any stub for this in `phpstub`, so I imagine there is some heavier logic used for this method, which is why I resorted to creating an issue instead :)

**Comment by weirdan** [comment]
`array_map()` is fine. You just did not provide it enough information on the input type because you confused `@param` with `@var`: https://psalm.dev/r/fc723b3432

**Comment by wouterj** [comment]
Thanks for your quick response @weirdan.

Seems indeed that my minimal example is wrong. I did however find this error when working on the minimal example for #4812. If I modify the (working) code in that issue from `array` to `list`, I get the same issue: https://psalm.dev/r/2774bb6f09

However, maybe it's directly related to the other issue instead of `array_map`?

**Comment by weirdan** [comment]
This seems to be specific to generic params. Simplified (with explicit generic and without recursion): https://psalm.dev/r/6a334b1371

@wouterj mind renaming the issue?

**Comment by muglug** [comment]
Yeah, I mentioned elsewhere (in the Symfony Slack) that this is a low-priority bug given it only affects templated types

**Comment by wouterj** [comment]
An even simpler example (without the mixed type): https://psalm.dev/r/6a334b1371

@weirdan thanks for looking into this. I've modified the title, but feel free to edit it. I'm not into Psalm internals, so have a hard time coming up with clear titles for the Psalm team :smiley:

**Comment by muglug** [comment]
`@template T of list<string>` is basically useless as a templated type, so I'm not sure this example is great

**Comment by weirdan** [comment]
> `@template T of list<string>` is basically useless as a templated type

I'd imagine it could be instantiated with `list<numeric-string>`, so should be still useful. Not in this context, just generally.

**Comment by orklah** [comment]
If I understood correctly, the final goal is this:
https://psalm.dev/r/2774bb6f09

The fact that it is transformed into templates under the hood is not relevant

**Comment by kkmuffme** [comment]
This is (and has always been) working correctly, the problem here was that @psalm-var instead of @psalm-param was used incorrectly, see: https://psalm.dev/r/7e9226b53a
